h/h_api/bulk_api/command_processor.py

```diff
@@ -0,0 +1,186 @@
+
+        self._report_back()
+
+    def _process_single_command(self, command):
+        """Process a single command."""
+
+        self.observer.observe_command(command, status=CommandStatus.AS_RECEIVED)
+
+        self.command_count += 1
+        self._check_command_count()
+
+        if isinstance(command, ConfigCommand):
+            self._configure(command.body)
+        else:
+            self._add_to_batch(command)
+
+        self.observer.observe_command(command, status=CommandStatus.POST_EXECUTE)
+
+    def _configure(self, config):
+        """Configure this object and the executor."""
+
+        if self.config is not None:
+            raise CommandSequenceError("Cannot currently re-configure jobs")
+
+        self.executor.configure(config)
+        self.config = config
+
+    def _add_to_batch(self, command):
+        """Add a single command to the batch.
+
+        When the batch size reaches maximum, or the task switches this will
+        trigger the commands to be flushed to the executor.
+        """
+
+        if self.config is None:
+            raise CommandSequenceError("Not configured yet")
+
+        with self.batch.add(command):
+            # If we have any id references like `"id": {"$ref": ...}` we need
+            # to fill these out before we pass them to the executor. We do this
+            # now to get the earliest warning if we have any id references
+            # which don't match up
+            self.id_refs.fill_out(command.body.raw)
+
+    def _check_command_count(self, final=False):
+        """Check the command count matches expectations.
+
+        :param final: This is the final count check, not incremental
+        """
+
+        if not self.config:
+            return
+
+        total = self.config.total_instructions
+
+        if not final:
+            if self.command_count > total:
+                raise InvalidDeclarationError(
+                    f"More instructions ({self.command_count}) received than declared ({total})"
+                )
+
+        elif self.command_count != total:
+            raise InvalidDeclarationError(
+                f"Expected more instructions. Found {self.command_count} expected {total}"
+            )
+
+    def _execute_batch(self, command_type, data_type, batch):
+        """Prepare and execute a batch of commands.
+
+        This is passed to the `CommandBatch` object which will call us back
+        here when a batch is ready.
+        """
+
+        # Get configuration for this combo of command and data type
+        default_config = self.config.defaults_for(command_type, data_type)
+
+        # Prep commands to be sent to the executor
+        #
+        # All items in batch are the same type. We can use the first one
+        # to process the items in place. This will effect any command type
+        # specific commands which can be done for the executor
+        batch[0].prepare_for_execute(batch, default_config)
```

Comment:
I had varitions of `batch[0].__class__` etc to make the point, but:

 * It's actually not any different
 * It makes it a pain to mock